File: src/main/java/at/hrechny/predictionsbot/controller/TelegramWebAppController.java
package at.hrechny.predictionsbot.controller;

import at.hrechny.predictionsbot.database.entity.MatchEntity;
import at.hrechny.predictionsbot.database.entity.RoundEntity;
import at.hrechny.predictionsbot.database.entity.UserEntity;
import at.hrechny.predictionsbot.database.model.MatchStatus;
import at.hrechny.predictionsbot.config.MessageResolver;
import at.hrechny.predictionsbot.exception.InputValidationException;
import at.hrechny.predictionsbot.exception.LimitExceededException;
import at.hrechny.predictionsbot.exception.interceptor.EnableErrorReport;
import at.hrechny.predictionsbot.model.LeagueRequest;
import at.hrechny.predictionsbot.model.LeagueResponse;
import at.hrechny.predictionsbot.model.Result;
import at.hrechny.predictionsbot.service.predictor.CompetitionService;
import at.hrechny.predictionsbot.service.predictor.LeagueService;
import at.hrechny.predictionsbot.service.predictor.PredictionService;
import at.hrechny.predictionsbot.service.predictor.UserService;
import at.hrechny.predictionsbot.util.HashUtils;
import at.hrechny.predictionsbot.util.ObjectUtils;
import java.util.Arrays;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;
import org.apache.commons.collections4.CollectionUtils;
import io.micronaut.context.annotation.Value;
import io.micronaut.core.annotation.Nullable;
import io.micronaut.http.HttpStatus;
import io.micronaut.http.MediaType;
import io.micronaut.http.HttpResponse;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Delete;
import io.micronaut.http.annotation.Get;
import io.micronaut.http.annotation.PathVariable;
import io.micronaut.http.annotation.Post;
import io.micronaut.http.annotation.Put;
import io.micronaut.http.annotation.Body;
import io.micronaut.http.annotation.QueryValue;
import io.micronaut.views.ModelAndView;

@Controller
@EnableErrorReport
public class TelegramWebAppController {

  @Value("${application.url}")
  private String applicationUrl;

  private final CompetitionService competitionService;
  private final PredictionService predictionService;
  private final LeagueService leagueService;
  private final UserService userService;
  private final HashUtils hashUtils;
  private final MessageResolver messageResolver;

  public TelegramWebAppController(
      CompetitionService competitionService,
      PredictionService predictionService,
      LeagueService leagueService,
      UserService userService,
      HashUtils hashUtils,
      MessageResolver messageResolver) {

    this.competitionService = competitionService;
    this.predictionService = predictionService;
    this.leagueService = leagueService;
    this.userService = userService;
    this.hashUtils = hashUtils;
    this.messageResolver = messageResolver;
  }

  @Get(value = "/webapp/{hash}/users/{userId}/predictions", produces = MediaType.TEXT_HTML)
  public ModelAndView<Map<String, Object>> getPredictions(
      @PathVariable("hash") String hash,
      @PathVariable("userId") Long userId,
      @QueryValue("competitionId") UUID competitionId,
      @Nullable @QueryValue("round") Integer roundNumber) {

    var user = userService.getUser(userId);

    RoundEntity round;
    if (roundNumber == null || roundNumber.equals(0)) {
      round = competitionService.getUpcomingRound(competitionId);
    } else {
      round = competitionService.getRound(competitionId, roundNumber);
    }

    if (round == null || CollectionUtils.isEmpty(round.getMatches())) {
      var model = webModel(user);
      model.put("competitionName", competitionService.getCompetition(competitionId).getName());
      return new ModelAndView<>("no-upcoming-matches", model);
    }

    var rounds = round.getSeason().getRounds().stream()
        .filter(roundEntity -> !roundEntity.getMatches().isEmpty())
        .filter(roundEntity -> roundEntity.getOrderNumber() != 0)
        .filter(ObjectUtils.distinctByKey(RoundEntity::getOrderNumber))
        .sorted(Comparator.comparingInt(RoundEntity::getOrderNumber))
        .toList();
    var fixtures = round.getSeason().getRounds().stream()
        .filter(roundEntity -> roundEntity.getOrderNumber() == round.getOrderNumber())
        .flatMap(roundEntity -> roundEntity.getMatches().stream())
        .sorted(Comparator.comparing(MatchEntity::getStartTime, Comparator.nullsLast(Comparator.naturalOrder())))
        .toList();

    var model = webModel(user);
    model.put("user", user);
    model.put("fixtures", fixtures);
    model.put("rounds", rounds);
    model.put("baseUrl", buildBaseUrl("predictions", userId, competitionId, null));
    return new ModelAndView<>("predictions", model);
  }

  @Get(value = "/webapp/{hash}/users/{userId}/results", produces = MediaType.TEXT_HTML)
  public ModelAndView<Map<String, Object>> getResults(
      @PathVariable("hash") String hash,
      @PathVariable("userId") Long userId,
      @QueryValue("competitionId") UUID competitionId,
      @Nullable @QueryValue("seasonId") UUID seasonId,
      @Nullable @QueryValue("round") Integer roundNumber) {

    var user = userService.getUser(userId);

    var season = seasonId != null ? competitionService.getSeason(seasonId) : competitionService.getCurrentSeason(competitionId);
    competitionService.refreshActiveFixtures(season.getId());

    List<Result> results;
    List<MatchEntity> matches;
    if (roundNumber != null && !roundNumber.equals(0)) {
      matches = season.getRounds().stream()
          .filter(roundEntity -> roundNumber.equals(roundEntity.getOrderNumber()))
          .flatMap(roundEntity -> roundEntity.getMatches().stream())
          .filter(match -> Arrays.asList(MatchStatus.STARTED, MatchStatus.FINISHED).contains(match.getStatus()))
          .filter(match -> CollectionUtils.isNotEmpty(match.getPredictions()))
          .sorted(Comparator.comparing(MatchEntity::getStartTime, Comparator.nullsLast(Comparator.naturalOrder())))
          .toList();
      results = predictionService.getResults(matches);
    } else {
      results = predictionService.getResults(season.getId());
      matches = season.getRounds().stream()
          .flatMap(roundEntity -> roundEntity.getMatches().stream())
          .filter(match -> Arrays.asList(MatchStatus.STARTED, MatchStatus.FINISHED).contains(match.getStatus()))
          .filter(match -> CollectionUtils.isNotEmpty(match.getPredictions()))
          .sorted(Comparator.comparing(MatchEntity::getStartTime, Comparator.nullsLast(Comparator.reverseOrder())))
          .limit(10)
          .toList();
    }

    var matchResults = matches.stream()
        .collect(Collectors.toMap(match -> match.getId().toString(), match -> predictionService.getResults(List.of(match))));
    var rounds = season.getRounds().stream()
        .flatMap(roundEntity -> roundEntity.getMatches().stream())
        .filter(match -> Arrays.asList(MatchStatus.STARTED, MatchStatus.FINISHED).contains(match.getStatus()))
        .filter(match -> CollectionUtils.isNotEmpty(match.getPredictions()))
        .map(MatchEntity::getRound)
        .filter(ObjectUtils.distinctByKey(RoundEntity::getOrderNumber))
        .sorted(Comparator.comparingInt(RoundEntity::getOrderNumber))
        .distinct()
        .toList();

    if (CollectionUtils.isEmpty(rounds)) {
      var model = webModel(user);
      model.put("competitionName", competitionService.getCompetition(competitionId).getName());
      return new ModelAndView<>("no-results", model);
    }

    var model = webModel(user);
    model.put("user", user);
    model.put("results", results);
    model.put("rounds", rounds);
    model.put("activeRound", roundNumber);
    model.put("matches", matches);
    model.put("matchResults", matchResults);
    model.put("competitionName", competitionService.getCompetition(competitionId).getName());
    model.put("baseUrl", buildBaseUrl("results", userId, competitionId, seasonId));

    return new ModelAndView<>("results", model);
  }

  @Get(value = "/webapp/{hash}/users/{userId}/leagues", produces = MediaType.TEXT_HTML)
  public ModelAndView<Map<String, Object>> getLeagues(
      @PathVariable("hash") String hash,
      @PathVariable("userId") Long userId) {

    var user = userService.getUser(userId);

    var model = webModel(user);
    model.put("user", user);
    return new ModelAndView<>("leagues", model);
  }

  @Post(value = "/webapp/{hash}/users/{userId}/leagues", produces = MediaType.APPLICATION_JSON)
  public HttpResponse<LeagueResponse> createLeague(
      @PathVariable("hash") String hash,
      @PathVariable("userId") Long userId,
      @Body LeagueRequest leagueRequest) {

    try {
      return HttpResponse.ok(leagueService.create(userId, leagueRequest));
    } catch (InputValidationException inputValidationException) {
      return HttpResponse.badRequest();
    } catch (LimitExceededException limitExceededException) {
      return HttpResponse.status(HttpStatus.CONFLICT);
    }
  }

  @Put(value = "/webapp/{hash}/users/{userId}/leagues/{leagueId}", produces = MediaType.APPLICATION_JSON)
  public HttpResponse<LeagueResponse> updateLeague(
      @PathVariable("hash") String hash,
      @PathVariable("userId") Long userId,
      @PathVariable("leagueId") UUID leagueId,
      @Body LeagueRequest leagueRequest) {

    try {
      return HttpResponse.ok(leagueService.update(userId, leagueId, leagueRequest));
    } catch (InputValidationException inputValidationException) {
      return HttpResponse.badRequest();
    }
  }

  @Post(value = "/webapp/{hash}/users/{userId}/leagues/{leagueId}", produces = MediaType.APPLICATION_JSON)
  public HttpResponse<LeagueResponse> joinLeague(
      @PathVariable("hash") String hash,
      @PathVariable("userId") Long userId,
      @PathVariable("leagueId") UUID leagueId) {

    try {
      return HttpResponse.ok(leagueService.join(userId, leagueId));
    } catch (InputValidationException inputValidationException) {
      return HttpResponse.badRequest();
    }
  }

  @Delete(value = "/webapp/{hash}/users/{userId}/leagues/{leagueId}", produces = MediaType.APPLICATION_JSON)
  public HttpResponse<LeagueResponse> deleteLeague(
      @PathVariable("hash") String hash,
      @PathVariable("userId") Long userId,
      @PathVariable("leagueId") UUID leagueId) {

    try {
      return HttpResponse.ok(leagueService.delete(userId, leagueId));
    } catch (InputValidationException inputValidationException) {
      return HttpResponse.badRequest();
    }
  }

  private String buildBaseUrl(String key, Long userId, UUID competitionId, UUID seasonId) {
    var url = applicationUrl + "/webapp/" + hashUtils.getHash(userId.toString()) + "/users/" + userId + "/" + key + "?competitionId=" + competitionId;
    if (seasonId != null) {
      url += "&seasonId=" + seasonId;
    }
    return url + "&round=";
  }

  private Map<String, Object> webModel(UserEntity user) {
    var model = new HashMap<String, Object>();
    model.put("i18n", messageResolver.forLocale(getLocale(user)));
    return model;
  }

  private Locale getLocale(UserEntity user) {
    return user.getLanguage() != null ? user.getLanguage() : user.getInitialLanguage();
  }
}

File: src/main/kotlin/at/hrechny/predictionsbot/controller/TelegramWebAppController.kt
package at.hrechny.predictionsbot.controller

import at.hrechny.predictionsbot.config.MessageResolver
import at.hrechny.predictionsbot.database.entity.MatchEntity
import at.hrechny.predictionsbot.database.entity.RoundEntity
import at.hrechny.predictionsbot.database.entity.UserEntity
import at.hrechny.predictionsbot.database.model.MatchStatus
import at.hrechny.predictionsbot.exception.InputValidationException
import at.hrechny.predictionsbot.exception.LimitExceededException
import at.hrechny.predictionsbot.exception.interceptor.EnableErrorReport
import at.hrechny.predictionsbot.model.LeagueRequest
import at.hrechny.predictionsbot.model.LeagueResponse
import at.hrechny.predictionsbot.model.Result
import at.hrechny.predictionsbot.service.predictor.CompetitionService
import at.hrechny.predictionsbot.service.predictor.LeagueService
import at.hrechny.predictionsbot.service.predictor.PredictionService
import at.hrechny.predictionsbot.service.predictor.UserService
import at.hrechny.predictionsbot.util.HashUtils
import io.micronaut.context.annotation.Value
import io.micronaut.core.annotation.Nullable
import io.micronaut.http.HttpResponse
import io.micronaut.http.HttpStatus
import io.micronaut.http.MediaType
import io.micronaut.http.annotation.Body
import io.micronaut.http.annotation.Controller
import io.micronaut.http.annotation.Delete
import io.micronaut.http.annotation.Get
import io.micronaut.http.annotation.PathVariable
import io.micronaut.http.annotation.Post
import io.micronaut.http.annotation.Put
import io.micronaut.http.annotation.QueryValue
import io.micronaut.views.ModelAndView
import org.apache.commons.collections4.CollectionUtils
import java.util.Comparator
import java.util.Locale
import java.util.UUID

@Controller
@EnableErrorReport
open class TelegramWebAppController(
    private val competitionService: CompetitionService,
    private val predictionService: PredictionService,
    private val leagueService: LeagueService,
    private val userService: UserService,
    private val hashUtils: HashUtils,
    private val messageResolver: MessageResolver,
) {
    @Value("\${application.url}")
    private lateinit var applicationUrl: String

    @Get(value = "/webapp/{hash}/users/{userId}/predictions", produces = [MediaType.TEXT_HTML])
    open fun getPredictions(
        @PathVariable("hash") hash: String,
        @PathVariable("userId") userId: Long,
        @QueryValue("competitionId") competitionId: UUID,
        @Nullable @QueryValue("round") roundNumber: Int?,
    ): ModelAndView<Map<String, Any>> {
        val user = userService.getUser(userId)
        val round = if (roundNumber == null || roundNumber == 0) {
            competitionService.getUpcomingRound(competitionId)
        } else {
            competitionService.getRound(competitionId, roundNumber)
        }

        if (round == null || CollectionUtils.isEmpty(round.matches)) {
            val model = webModel(user)
            model["competitionName"] = competitionService.getCompetition(competitionId).name!!
            return ModelAndView("no-upcoming-matches", model)
        }

        val season = round.season!!
        val rounds = season.rounds
            .filter { roundEntity -> roundEntity.matches.isNotEmpty() }
            .filter { roundEntity -> roundEntity.orderNumber != 0 }
            .distinctBy(RoundEntity::orderNumber)
            .sortedBy(RoundEntity::orderNumber)
        val fixtures = season.rounds
            .filter { roundEntity -> roundEntity.orderNumber == round.orderNumber }
            .flatMap { roundEntity -> roundEntity.matches }
            .sortedWith(compareBy(Comparator.nullsLast(Comparator.naturalOrder())) { match -> match.startTime })

        val model = webModel(user)
        model["user"] = user
        model["fixtures"] = fixtures
        model["rounds"] = rounds
        model["baseUrl"] = buildBaseUrl("predictions", userId, competitionId, null)
        return ModelAndView("predictions", model)
    }

    @Get(value = "/webapp/{hash}/users/{userId}/results", produces = [MediaType.TEXT_HTML])
    open fun getResults(
        @PathVariable("hash") hash: String,
        @PathVariable("userId") userId: Long,
        @QueryValue("competitionId") competitionId: UUID,
        @Nullable @QueryValue("seasonId") seasonId: UUID?,
        @Nullable @QueryValue("round") roundNumber: Int?,
    ): ModelAndView<Map<String, Any>> {
        val user = userService.getUser(userId)
        val season = if (seasonId != null) {
            competitionService.getSeason(seasonId)
        } else {
            competitionService.getCurrentSeason(competitionId)
        }
        competitionService.refreshActiveFixtures(season.id!!)

        val results: List<Result>
        val matches: List<MatchEntity>
        if (roundNumber != null && roundNumber != 0) {
            matches = season.rounds
                .filter { roundEntity -> roundNumber == roundEntity.orderNumber }
                .flatMap { roundEntity -> roundEntity.matches }
                .filter { match -> match.status in setOf(MatchStatus.STARTED, MatchStatus.FINISHED) }
                .filter { match -> CollectionUtils.isNotEmpty(match.predictions) }
                .sortedWith(compareBy(Comparator.nullsLast(Comparator.naturalOrder())) { match -> match.startTime })
            results = predictionService.getResults(matches)
        } else {
            results = predictionService.getResults(season.id!!)
            matches = season.rounds
                .flatMap { roundEntity -> roundEntity.matches }
                .filter { match -> match.status in setOf(MatchStatus.STARTED, MatchStatus.FINISHED) }
                .filter { match -> CollectionUtils.isNotEmpty(match.predictions) }
                .sortedWith(compareBy(Comparator.nullsLast(Comparator.reverseOrder())) { match -> match.startTime })
                .take(10)
        }

        val matchResults = matches.associate { match ->
            match.id.toString() to predictionService.getResults(listOf(match))
        }
        val rounds = season.rounds
            .flatMap { roundEntity -> roundEntity.matches }
            .filter { match -> match.status in setOf(MatchStatus.STARTED, MatchStatus.FINISHED) }
            .filter { match -> CollectionUtils.isNotEmpty(match.predictions) }
            .map { match -> match.round!! }
            .distinctBy(RoundEntity::orderNumber)
            .sortedBy(RoundEntity::orderNumber)
            .distinct()

        if (CollectionUtils.isEmpty(rounds)) {
            val model = webModel(user)
            model["competitionName"] = competitionService.getCompetition(competitionId).name!!
            return ModelAndView("no-results", model)
        }

        val model = webModel(user)
        model["user"] = user
        model["results"] = results
        model["rounds"] = rounds
        if (roundNumber != null) {
            model["activeRound"] = roundNumber
        }
        model["matches"] = matches
        model["matchResults"] = matchResults
        model["competitionName"] = competitionService.getCompetition(competitionId).name!!
        model["baseUrl"] = buildBaseUrl("results", userId, competitionId, seasonId)

        return ModelAndView("results", model)
    }

    @Get(value = "/webapp/{hash}/users/{userId}/leagues", produces = [MediaType.TEXT_HTML])
    open fun getLeagues(
        @PathVariable("hash") hash: String,
        @PathVariable("userId") userId: Long,
    ): ModelAndView<Map<String, Any>> {
        val user = userService.getUser(userId)
        val model = webModel(user)
        model["user"] = user
        return ModelAndView("leagues", model)
    }

    @Post(value = "/webapp/{hash}/users/{userId}/leagues", produces = [MediaType.APPLICATION_JSON])
    open fun createLeague(
        @PathVariable("hash") hash: String,
        @PathVariable("userId") userId: Long,
        @Body leagueRequest: LeagueRequest,
    ): HttpResponse<LeagueResponse> =
        try {
            HttpResponse.ok(leagueService.create(userId, leagueRequest))
        } catch (inputValidationException: InputValidationException) {
            HttpResponse.badRequest()
        } catch (limitExceededException: LimitExceededException) {
            HttpResponse.status(HttpStatus.CONFLICT)
        }

    @Put(value = "/webapp/{hash}/users/{userId}/leagues/{leagueId}", produces = [MediaType.APPLICATION_JSON])
    open fun updateLeague(
        @PathVariable("hash") hash: String,
        @PathVariable("userId") userId: Long,
        @PathVariable("leagueId") leagueId: UUID,
        @Body leagueRequest: LeagueRequest,
    ): HttpResponse<LeagueResponse> =
        try {
            val response = leagueService.update(userId, leagueId, leagueRequest)
            if (response != null) HttpResponse.ok(response) else HttpResponse.ok()
        } catch (inputValidationException: InputValidationException) {
            HttpResponse.badRequest()
        }

    @Post(value = "/webapp/{hash}/users/{userId}/leagues/{leagueId}", produces = [MediaType.APPLICATION_JSON])
    open fun joinLeague(
        @PathVariable("hash") hash: String,
        @PathVariable("userId") userId: Long,
        @PathVariable("leagueId") leagueId: UUID,
    ): HttpResponse<LeagueResponse> =
        try {
            val response = leagueService.join(userId, leagueId)
            if (response != null) HttpResponse.ok(response) else HttpResponse.ok()
        } catch (inputValidationException: InputValidationException) {
            HttpResponse.badRequest()
        }

    @Delete(value = "/webapp/{hash}/users/{userId}/leagues/{leagueId}", produces = [MediaType.APPLICATION_JSON])
    open fun deleteLeague(
        @PathVariable("hash") hash: String,
        @PathVariable("userId") userId: Long,
        @PathVariable("leagueId") leagueId: UUID,
    ): HttpResponse<LeagueResponse> =
        try {
            val response = leagueService.delete(userId, leagueId)
            if (response != null) HttpResponse.ok(response) else HttpResponse.ok()
        } catch (inputValidationException: InputValidationException) {
            HttpResponse.badRequest()
        }

    private fun buildBaseUrl(key: String, userId: Long, competitionId: UUID, seasonId: UUID?): String {
        var url = "$applicationUrl/webapp/${hashUtils.getHash(userId.toString())}/users/$userId/$key?competitionId=$competitionId"
        if (seasonId != null) {
            url += "&seasonId=$seasonId"
        }
        return "$url&round="
    }

    private fun webModel(user: UserEntity): MutableMap<String, Any> =
        hashMapOf("i18n" to messageResolver.forLocale(getLocale(user)))

    private fun getLocale(user: UserEntity): Locale? = user.language ?: user.initialLanguage
}
